### ISSUE TITLE
KSY schema: generate human-readable documentation

### ISSUE THREAD

**Issue by GreyCat** [kaitai-io/kaitai_struct#655]
We're still stuck with http://doc.kaitai.io/ksy_reference.html, which is frankly horrible, while we have a new and shiny https://github.com/kaitai-io/ksy_schema

Any chance anyone knows if we can generate human-readable KSY reference from our KSY schema?

**Comment by GreyCat** [comment]
Ideas to explore:

* [jsonschema2md](https://github.com/adobe/jsonschema2md)
* [json-schema-for-humans](https://pypi.org/project/json-schema-for-humans/)
* https://www.liquid-technologies.com/json-schema-documentation
* [Docson](https://dzone.com/articles/publishing-json-schema-documentaton-with-docson)
* https://blog.cloudflare.com/cloudflares-json-powered-documentation-generator/

Meta question on SO: https://stackoverflow.com/questions/46747381/json-schema-documentation-generators

**Comment by KOLANICH** [comment]
As usual I vote for staying away of NPM and Node as much as possible :)

**Comment by GreyCat** [comment]
### [jsonschema2md](https://github.com/adobe/jsonschema2md)

Fails miserably, probably due to some unsupported elements of the schema:

```
[WARN] schema /home/greycat/git/ksy_schema/ksy_schema.json has no $id
finished reading /home/greycat/git/ksy_schema/ksy_schema.json, beginning processing...
unresolved reference: #/definitions/Identifier
unresolved reference: #/definitions/StringOrInteger
unresolved reference: #/definitions/StringOrInteger
unresolved reference: #/definitions/StringOrInteger
unresolved reference: #/definitions/Identifier
unresolved reference: #/definitions/StringOrInteger
unresolved reference: #/definitions/StringOrInteger
unresolved reference: #/definitions/StringOrInteger
unresolved reference: #/definitions/MetaSpec
unresolved reference: #/definitions/Doc
unresolved reference: #/definitions/DocRef
unresolved reference: #/definitions/ParamsSpec
unresolved reference: #/definitions/Attributes
unresolved reference: #/definitions/TypesSpec
unresolved reference: #/definitions/InstancesSpec
unresolved reference: #/definitions/EnumsSpec
unresolved reference: #/definitions/Doc
unresolved reference: #/definitions/DocRef
unresolved reference: #/definitions/EnumsSpec
unresolved reference: #/definitions/InstancesSpec
unresolved reference: #/definitions/MetaSpec
unresolved reference: #/definitions/ParamsSpec
unresolved reference: #/definitions/Attributes
unresolved reference: #/definitions/TypesSpec
[ERROR] Error occured in processing schema at path %s /home/greycat/git/ksy_schema/ksy_schema.json
[ERROR] { TypeError: /home/greycat/git/ksy_schema/node_modules/@adobe/jsonschema2md/templates/md/pattern-property.ejs:22
    20| 
    21| * <%=i18n.__('patternPropertyIsPropertyPattern')%>
 >> 22| * <%=i18n.__('patternPropertyType')%> <%=schema.simpletype %><% if (schema.simpletype.match(/\[\]\[\]/)||schema.simpletype==="`array[]`") { %> <%=i18n.__('patternPropertyNestedArray')%><% } %><% if (schema.type==='array') { -%>
    23| <%   if (schema.maxItems!==undefined&&schema.minItems!==undefined) { %>* <%=i18n.__('patternPropertyBetween')%> `<%=schema.minItems %>` <%=i18n.__('patternPropertyBetweenAnd')%> `<%=schema.maxItems %>` <%=i18n.__('nestedPropertyEnumKnownVaules')%><% }
    24| else if (schema.maxItems!==undefined                             ) { %>* <%=i18n.__('patternPropertyMaxItemsPrefix')%> `<%=schema.maxItems %>` <%=i18n.__('patternPropertyMaxItemsSuffix')%><% }
    25| else if (schema.minItems!==undefined                             ) { %>* <%=i18n.__('patternPropertyMinItemsPrefix')%> `<%=schema.minItems %>` <%=i18n.__('patternPropertyMinItemsSuffix')%><% } %>

Cannot read property 'match' of undefined
```

### [json-schema-for-humans](https://pypi.org/project/json-schema-for-humans/)

Generates .html file with *some* good content, but fails at `types`, `instances`, `enums` being documented just as "objects", i.e. like fails to understand `$ref`:

```json
    "TypesSpec": {
      "type": "object",
      "additionalProperties": false,
      "patternProperties": {
        "^([a-z0-9_])+$": {
          "$ref": "#/definitions/TypeSpec"
        }
      }
    },
```

Also, .html file is not exactly ideal for integration into https://doc.kaitai.io/ offering.

**Comment by GreyCat** [comment]
## [prmd](https://github.com/interagent/prmd)

* Fails `prmd verify` with the following, so likely it does not support draft-07:

```
ksy_schema.json: Unknown $schema: http://json-schema.org/draft-07/schema#.
```

* `prmd render --template schemata.erb` (which is supposed to generate Markdown docs) fails with:

```
Error in resource: doc
Traceback (most recent call last):
        15: from /usr/local/bin/prmd:23:in `<main>'
        14: from /usr/local/bin/prmd:23:in `load'
        13: from /var/lib/gems/2.5.0/gems/prmd-0.13.0/bin/prmd:6:in `<top (required)>'
        12: from /var/lib/gems/2.5.0/gems/prmd-0.13.0/lib/prmd/cli.rb:99:in `run'
        11: from /var/lib/gems/2.5.0/gems/prmd-0.13.0/lib/prmd/cli/base.rb:141:in `run'
        10: from /var/lib/gems/2.5.0/gems/prmd-0.13.0/lib/prmd/cli/render.rb:47:in `execute'
         9: from /var/lib/gems/2.5.0/gems/prmd-0.13.0/lib/prmd/commands/render.rb:34:in `render'
         8: from /var/lib/gems/2.5.0/gems/prmd-0.13.0/lib/prmd/core/renderer.rb:50:in `render'
         7: from /var/lib/gems/2.5.0/gems/erubis-2.7.0/lib/erubis/evaluator.rb:65:in `result'
         6: from /var/lib/gems/2.5.0/gems/erubis-2.7.0/lib/erubis/evaluator.rb:65:in `eval'
         5: from /var/lib/gems/2.5.0/gems/prmd-0.13.0/lib/prmd/templates/schema.erb:5:in `result'
         4: from /var/lib/gems/2.5.0/gems/prmd-0.13.0/lib/prmd/templates/schema.erb:5:in `map'
         3: from /var/lib/gems/2.5.0/gems/prmd-0.13.0/lib/prmd/templates/schema.erb:9:in `block in result'
         2: from /var/lib/gems/2.5.0/gems/erubis-2.7.0/lib/erubis/evaluator.rb:65:in `result'
         1: from /var/lib/gems/2.5.0/gems/erubis-2.7.0/lib/erubis/evaluator.rb:65:in `eval'
/var/lib/gems/2.5.0/gems/prmd-0.13.0/lib/prmd/templates/schemata.md.erb:9:in `result': undefined method `split' for nil:NilClass (NoMethodError)
```

Alas, looks like we can't use it :(

**Comment by generalmimon** [comment]
## [Docson](https://github.com/lbovet/docson)

Looks pretty good, here you can examine the [**online demo with KSY schema loaded**](https://lbovet.github.io/docson/index.html#https://raw.githubusercontent.com/kaitai-io/ksy_schema/master/ksy_schema.json). (please be patient and don't be discouraged by the initial bare links, the rendering usually takes 10-15 seconds)

However, Docson runs completely in the browser, so the JSON schema parsing runs on the client side at run time. That's probably why the demo I linked renders so slow, but it possibly doesn't have to be the case, it's just about choosing the right JSON linking method. This example runs pretty fast: http://lbovet.github.io/docson/examples/example.html

Of course, it would be better to have some backend tool to produce some nice HTML output, but I spent several hours searching for it, tried a number of tools and I wasn't satisfied with the results as much as with Docson (the output was either difficult to read, or the implementation was missing some essential JSON schema features so that the output generation failed or was unusable). There seems to be a lot of JSON schema docs generators, but the vast majority of them just focus on HTTP REST APIs (with all the URL routes and HTTP methods `GET`, `PUT`, `DELETE` etc.), which are not quite  suitable for our use case.

In short, Docson is certainly the best tool I have found so far and I think nothing prevents deploying it in place of KSY reference. And we can also use Markdown syntax in the `description` of the properties (so it doesn't have to be just plain text), which can help the readability.

**Comment by dblanchette** [comment]
Hi,

I am the creator of [json-schema-for-humans](https://github.com/coveooss/json-schema-for-humans).

Since this analysis was done, the library has evolved quite a bit! It is now able to render schemas with recursive definitions (it detects those cases and replaces the definition with an anchor link to the original one). It will also not get confused by properties named the same as keywords anymore.

I have used your schema to test the library which helped me a lot to tackle different use cases I did not think about.

I work on this library on my free time because I enjoy it. I enjoy it even more if it can be useful to other people. I understand you already implemented docson and you do not necessarily want to change things up again, but if you have some time to test my lib again and provide feedback, that would be help me a lot.

Feel free to open issues in our repo if you have suggestions or bugs :)
